tools/build-image-if-needed.sh

```diff
@@ -11,7 +11,17 @@ fi
 BUILD_DIRECTORY=$1
 TAG_NAME=$2
 
-if git diff --exit-code remotes/origin/master...HEAD -- ${BUILD_DIRECTORY}; then
+BRANCH="$(git rev-parse --abbrev-ref HEAD)"
+
+# If this script is being run *on* the master branch, then we want to diff
+# with the previous commit on master. Otherwise, we diff with master itself.
+if [[ "$BRANCH" == "master" ]]; then
+  DIFF_SOURCE="HEAD^1"
+else
+  DIFF_SOURCE="remotes/origin/master"
+fi
+
+if git diff --exit-code $DIFF_SOURCE...HEAD -- ${BUILD_DIRECTORY}; then
```

Comment:
I still can't tell from the Git docs if `...` or `..` is actually desired here 🙃

Reply:
I believe `...` is what we want. Consider:
```
branch_origin ---- master_changes --- master
        |
        +------ branch_changes ---- HEAD
```

Assume that both `branch_changes` AND `master_changes` introduce changes to `BUILD_DIRECTORY`. If we use `..` then we will get a diff that _adds_ the changes from `branch_changes` and _removes_ the changes from `master_changes` (these are the changes needed to turn `master` into `HEAD`). On the other hand, if we use `...` then we _only_ get the changes in `branch_changes` (these are the changes needed to turn `branch_origin` into `HEAD`).

We only want to build the image if we've introduced changes in the current branch (we don't care about what `master` has done in the meantime) so we want `...`.

Does that sound correct to you?

Reply:
That sounds correct! I'll roll with that.